# MLOps Workshop: Modell-Engineering und Training

## Einführung
In diesem Notebook werden wir uns mit dem Modell-Engineering und Training für unser Customer Churn Prediction Projekt beschäftigen. Wir werden MLflow für das Experiment-Tracking verwenden und verschiedene Modelle evaluieren.

## Lernziele
Nach Abschluss dieses Notebooks werden Sie:
- Verstehen, wie man MLflow für Experiment-Tracking einsetzt
- Verschiedene Modelle mit unterschiedlichen Hyperparametern trainieren können
- Best Practices für Modell-Evaluierung kennen
- Wissen, wie man Modelle im MLflow Model Registry registriert

## 1. Setup und Daten laden


In [31]:
# Benötigte Bibliotheken importieren
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, roc_auc_score
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.ensemble import GradientBoostingClassifier
import mlflow
import mlflow.sklearn
import datetime
import matplotlib.pyplot as plt
import seaborn as sns

# MLflow Tracking lokal einrichten (ohne Server)
mlflow.set_tracking_uri("file:./mlruns")


In [3]:
# Daten laden und vorbereiten
processed_data = pd.read_csv('../data/processed/loan_data_preprocessed.csv')



In [4]:
# Train-Test-Split
X = processed_data.drop(columns="not.fully.paid")
y= processed_data["not.fully.paid"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

scaler = StandardScaler()
X_train= scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [5]:
# Modell erstellen

model = LogisticRegression(max_iter=1000, class_weight="balanced")
model.fit(X_train, y_train)

y_pred = model.predict(X_test)
y_prob = model.predict_proba(X_test)[:,1]

print("Accuracy:", accuracy_score(y_test, y_pred))
print("ROC-AUC:", roc_auc_score(y_test, y_prob))
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))


Accuracy: 0.6482254697286013
ROC-AUC: 0.7127535081560176
[[1033  578]
 [  96  209]]
              precision    recall  f1-score   support

           0       0.91      0.64      0.75      1611
           1       0.27      0.69      0.38       305

    accuracy                           0.65      1916
   macro avg       0.59      0.66      0.57      1916
weighted avg       0.81      0.65      0.69      1916



### Auswertung der Modellleistung

Das trainierte Modell erreicht eine Accuracy von 0.65 sowie eine ROC-AUC von 0.71. Die ROC-AUC zeigt, dass das Modell grundsätzlich in der Lage ist, zwischen vollständig zurückgezahlten Krediten und nicht vollständig zurückgezahlten Krediten zu unterscheiden. Ein Wert von über 0.7 deutet auf eine moderate Trennfähigkeit des Modells hin.

Die Confusion Matrix zeigt, dass von insgesamt 305 tatsächlich nicht vollständig zurückgezahlten Krediten 209 korrekt erkannt wurden, während 96 Ausfälle nicht erkannt wurden. Daraus ergibt sich ein Recall von 0.69 für die Klasse „not.fully.paid = 1“, was bedeutet, dass etwa **69 % der tatsächlichen Kreditausfälle identifiziert werden.

Die Precision für die Ausfallklasse beträgt 0.27. Das bedeutet, dass nur etwa 27 % der als riskant eingestuften Kredite tatsächlich nicht vollständig zurückgezahlt werden, während ein größerer Teil dieser Vorhersagen Fehlalarme darstellt. Dies wird auch in der Confusion Matrix sichtbar, in der 578 Kredite fälschlicherweise als riskant klassifiziert wurden.

Insgesamt zeigt das Modell damit eine relativ hohe Sensitivität gegenüber Kreditausfällen, erkennt also einen großen Teil der tatsächlich problematischen Kredite. Gleichzeitig entstehen jedoch viele False Positives, da das Modell auch Kredite als riskant einstuft, die letztlich vollständig zurückgezahlt werden.

Diese Ergebnisse verdeutlichen, dass das Modell grundsätzlich Muster im Datensatz erkennen kann, die mit Kreditausfällen zusammenhängen, die Vorhersagen jedoch weiterhin mit einer gewissen Unsicherheit verbunden sind.

## 2. MLflow Experiment erstellen

### Aufgabe 1
Erstellen Sie ein neues MLflow Experiment für unser Churn-Vorhersage-Projekt. Setzen Sie einen aussagekräftigen Namen und eine Beschreibung.

<details>
<summary>👉 Lösung anzeigen</summary>

```python
# MLflow Experiment erstellen
experiment_name = "customer_churn_prediction"
experiment = mlflow.set_experiment(experiment_name)

# Experiment-Details ausgeben
print(f"Experiment Name: {experiment_name}")
print(f"Experiment ID: {experiment.experiment_id}")
```
</details>

In [37]:
# MLflow-Setup
import mlflow
import mlflow.sklearn

mlflow.set_tracking_uri("file:../mlruns")
mlflow.set_experiment("Loan Default Prediction")

print(mlflow.get_tracking_uri())

file:../mlruns


In [38]:
def train_and_evaluate_model(model, X_train, X_test, y_train, y_test, model_name, params=None):
    """
    Trainiert ein Modell und trackt die Ergebnisse mit MLflow.
    """

    with mlflow.start_run(run_name=model_name) as run:
        # Parameter loggen
        if params:
            mlflow.log_params(params)

        # Modell trainieren
        model.fit(X_train, y_train)

        # Vorhersagen
        y_pred = model.predict(X_test)

        # Metriken berechnen
        accuracy = accuracy_score(y_test, y_pred)
        precision = precision_score(y_test, y_pred, pos_label=1, zero_division=0)
        recall = recall_score(y_test, y_pred, pos_label=1, zero_division=0)
        f1 = f1_score(y_test, y_pred, pos_label=1, zero_division=0)

        # Metriken loggen
        mlflow.log_metric("accuracy", accuracy)
        mlflow.log_metric("precision", precision)
        mlflow.log_metric("recall", recall)
        mlflow.log_metric("f1_score", f1)

        # ROC-AUC nur wenn predict_proba vorhanden ist
        roc_auc = None
        if hasattr(model, "predict_proba"):
            y_prob = model.predict_proba(X_test)[:, 1]
            roc_auc = roc_auc_score(y_test, y_prob)
            mlflow.log_metric("roc_auc", roc_auc)

        # Modell loggen
        mlflow.sklearn.log_model(model, name="model")

        return run.info.run_id, {
            "accuracy": accuracy,
            "precision": precision,
            "recall": recall,
            "f1_score": f1,
            "roc_auc": roc_auc
        }

In [45]:
# Logistic Regression 
#Hyperparameter für Logistic Regression
logreg_params_list = [
    {
        "C": 0.1,
        "max_iter": 1000,
        "class_weight": None,
        "random_state": 42
    },
    {
        "C": 1.0,
        "max_iter": 1000,
        "class_weight": "balanced",
        "random_state": 42
    },
    {
        "C": 10.0,
        "max_iter": 1000,
        "class_weight": "balanced",
        "random_state": 42
    }
]

In [46]:
from sklearn.linear_model import LogisticRegression

for params in logreg_params_list:

    model = LogisticRegression(**params)

    run_id, metrics = train_and_evaluate_model(
        model,
        X_train,
        X_test,
        y_train,
        y_test,
        model_name="LogisticRegression",
        params=params
    )

    results.append({
        "run_id": run_id,
        "params": params,
        "metrics": metrics,
        "model_name": "LogisticRegression"
    })

2026/03/10 14:21:32 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
2026/03/10 14:21:35 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
2026/03/10 14:21:38 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mec

In [ ]:
# Ramdom Forest  

# Hyperparameter für Random Forest
rf_params_list = [
    {
        'n_estimators': 100,
        'max_depth': 10,
        'min_samples_split': 2,
        'random_state': 42
    },
    {
        'n_estimators': 200,
        'max_depth': 15,
        'min_samples_split': 5,
        'random_state': 42
    },
    {
        'n_estimators': 300,
        'max_depth': None,
        'min_samples_split': 10,
        'random_state': 42
    }
]



In [41]:
# Trainiere und logge Random Forest Modelle mit verschiedenen Hyperparametern
results = []

for params in rf_params_list:

    model = RandomForestClassifier(**params)

    run_id, metrics = train_and_evaluate_model(
        model,
        X_train,
        X_test,
        y_train,
        y_test,
        model_name="RandomForest",
        params=params
    )

    results.append({
        "run_id": run_id,
        "params": params,
        "metrics": metrics,
        "model_name": "RandomForest"
    })

2026/03/10 14:15:30 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
2026/03/10 14:15:35 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
2026/03/10 14:15:42 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mec

In [42]:
# Gradient Boosting Modell 
# Hyperparameter für Gradient Boosting

gb_params_list = [
    {
        "n_estimators": 100,
        "learning_rate": 0.05,
        "max_depth": 3,
        "random_state": 42
    },
    {
        "n_estimators": 200,
        "learning_rate": 0.05,
        "max_depth": 3,
        "random_state": 42
    },
    {
        "n_estimators": 100,
        "learning_rate": 0.1,
        "max_depth": 4,
        "random_state": 42
    }
]

In [43]:
gb_results = []

for params in gb_params_list:

    model = GradientBoostingClassifier(**params)

    run_id, metrics = train_and_evaluate_model(
        model,
        X_train,
        X_test,
        y_train,
        y_test,
        model_name="GradientBoosting",
        params=params
    )

    gb_results.append({
        "run_id": run_id,
        "params": params,
        "metrics": metrics,
        "model_name": "GradientBoosting"
    })

2026/03/10 14:15:51 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
2026/03/10 14:15:58 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
2026/03/10 14:16:03 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mec

In [47]:
# Bestes Modell anhand des ROC AUC Scores auswählen uswählen
best_run = max(results, key=lambda x: x['metrics']['f1_score'])
best_run_id = best_run['run_id']

# Modell im Registry registrieren
model_name = "loan_default_predictor"

model_version = mlflow.register_model(
    f"runs:/{best_run_id}/model",
    model_name
)

print(f"Model '{model_name}' wurde als Version {model_version.version} registriert")
print("Bestes Modell:")
print(best_run["params"])

print("Metriken:")
print(best_run["metrics"])


Registered model 'loan_default_predictor' already exists. Creating a new version of this model...
2026/03/10 14:21:51 WARNING mlflow.tracking._model_registry.fluent: Run with id e5ae368262f04f258b6cde49c40a4d5d has no artifacts at artifact path 'model', registering model based on models:/m-9107776bcec043d1be27088089c87ada instead


Model 'loan_default_predictor' wurde als Version 4 registriert
Bestes Modell:
{'C': 10.0, 'max_iter': 1000, 'class_weight': 'balanced', 'random_state': 42}
Metriken:
{'accuracy': 0.6477035490605428, 'precision': 0.26641414141414144, 'recall': 0.6918032786885245, 'f1_score': 0.3846855059252507, 'roc_auc': 0.7128756194604715}


Created version '4' of model 'loan_default_predictor'.


In [48]:
from mlflow.tracking import MlflowClient
import datetime

client = MlflowClient()

# Modell auf Staging setzen
client.transition_model_version_stage(
    name="loan_default_predictor",
    version=model_version.version,
    stage="Staging",
    archive_existing_versions=True
)

# Modellbeschreibung hinzufügen
client.update_model_version(
    name="loan_default_predictor",
    version=model_version.version,
    description=f"""
Loan Default Prediction Model v{model_version.version}

- Modelltyp: Random Forest
- Training durchgeführt am: {datetime.datetime.now().strftime('%Y-%m-%d %H:%M')}
- Beste Parameter: {best_run['params']}

Performance:
- Accuracy: {best_run['metrics']['accuracy']:.4f}
- Precision: {best_run['metrics']['precision']:.4f}
- Recall: {best_run['metrics']['recall']:.4f}
- F1-Score: {best_run['metrics']['f1_score']:.4f}
- ROC-AUC: {best_run['metrics']['roc_auc']:.4f}
"""
)

C:\Users\bjoer\AppData\Local\Temp\ipykernel_10320\4255999107.py:7: FutureWarning: ``mlflow.tracking.client.MlflowClient.transition_model_version_stage`` is deprecated since 2.9.0. Model registry stages will be removed in a future major release. To learn more about the deprecation of model registry stages, see our migration guide here: https://mlflow.org/docs/latest/model-registry.html#migrating-from-stages
  client.transition_model_version_stage(


<ModelVersion: aliases=[], creation_timestamp=1773148911437, current_stage='Staging', deployment_job_state=None, description=('\n'
 'Loan Default Prediction Model v4\n'
 '\n'
 '- Modelltyp: Random Forest\n'
 '- Training durchgeführt am: 2026-03-10 14:26\n'
 "- Beste Parameter: {'C': 10.0, 'max_iter': 1000, 'class_weight': 'balanced', "
 "'random_state': 42}\n"
 '\n'
 'Performance:\n'
 '- Accuracy: 0.6477\n'
 '- Precision: 0.2664\n'
 '- Recall: 0.6918\n'
 '- F1-Score: 0.3847\n'
 '- ROC-AUC: 0.7129\n'), last_updated_timestamp=1773149169454, metrics=[<Metric: dataset_digest=None, dataset_name=None, key='accuracy', model_id='m-9107776bcec043d1be27088089c87ada', run_id='e5ae368262f04f258b6cde49c40a4d5d', step=0, timestamp=1773148898463, value=0.6477035490605428>,
 <Metric: dataset_digest=None, dataset_name=None, key='f1_score', model_id='m-9107776bcec043d1be27088089c87ada', run_id='e5ae368262f04f258b6cde49c40a4d5d', step=0, timestamp=1773148898478, value=0.3846855059252507>,
 <Metric: datas


## Hausaufgaben und weiterführende Übungen
1. Experimentieren Sie mit weiteren Modelltypen (z.B. XGBoost, LightGBM)
2. Implementieren Sie eine Cross-Validation in den Training-Prozess
3. Erweitern Sie das Feature Importance Plot um weitere Visualisierungen
4. Fügen Sie Custom Tags zu den MLflow Runs hinzu
5. Implementieren Sie eine automatische Modellversionierung basierend auf Performance-Metriken

## Nächste Schritte
- Fahren Sie mit dem Deployment-Notebook fort
- Lernen Sie, wie man das Modell mit FastAPI bereitstellt
- Implementieren Sie Monitoring für das deployte Modell

## Hilfreiche MLflow-Befehle
```bash
# MLflow UI starten
mlflow ui

# Experiments auflisten
mlflow experiments list

# Run Details anzeigen
mlflow runs describe <run-id>

# Modell laden
mlflow models load runs:/<run-id>/model
```